Mathlib/Topology/UniformSpace/Ascoli.lean

```diff
@@ -0,0 +1,483 @@
+/-
+Copyright (c) 2022 Anatole Dedecker. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anatole Dedecker
+-/
+import Mathlib.Topology.UniformSpace.Equicontinuity
+import Mathlib.Topology.UniformSpace.Equiv
+
+/-!
+# Ascoli Theorem
+
+In this file, we prove the general **Arzela-Ascoli theorem**, and various related statements about
+the topology of equicontinuous subsetes of `X →ᵤ[𝔖] α`, where `X` is a topological space, `𝔖` is
+a family of compact subsets of `X`, and `α` is a uniform space.
+
+## Main statements
+
+* If `X` is a compact space, then the uniform structures of uniform convergence and pointwise
+  convergence coincide on equicontinuous subsets. This is the key fact that makes equicontinuity
+  important in functional analysis. We state various versions of it:
+  - as an equality of `UniformSpace`s: `Equicontinuous.comap_uniformFun_eq`
+  - in terms of `UniformInducing`: `Equicontinuous.uniformInducing_uniformFun_iff_pi`
+  - in terms of `Inducing`: `Equicontinuous.inducing_uniformFun_iff_pi`
+  - in terms of convergence along a filter: `Equicontinuous.tendsto_uniformFun_iff_pi`
+* As a consequence, if `𝔖` is a family of compact subsets of `X`, then the uniform structures of
+  uniform convergence on `𝔖` and pointwise convergence on `⋃₀ 𝔖` coincide on equicontinuous
+  subsets. Again, we prove multiple variations:
+  - as an equality of `UniformSpace`s: `EquicontinuousOn.comap_uniformOnFun_eq`
+  - in terms of `UniformInducing`: `EquicontinuousOn.uniformInducing_uniformOnFun_iff_pi'`
+  - in terms of `Inducing`: `EquicontinuousOn.inducing_uniformOnFun_iff_pi'`
+  - in terms of convergence along a filter: `EquicontinuousOn.tendsto_uniformOnFun_iff_pi'`
+* The **Arzela-Ascoli theorem** follows from the previous fact and Tykhonov's theorem.
+  All of its variations can be found under the `ArzelaAscoli` namespace.
+
+## Implementation details
+
+* The statements in this file may be a bit daunting because we prove everything for families and
+  embeddings instead of subspaces with the subspace topology. This is done because, in practice,
+  one would rarely work with `X →ᵤ[𝔖] α` directly, so we need to provide API for bringing back the
+  statements to various other types, such as `C(X, Y)` or `E →L[𝕜] F`. To counteract this, all
+  statements (as well as most proofs!) are documented quite thouroughly.
+
+* A lot of statements assume `∀ K ∈ 𝔖, EquicontinuousOn F K` instead of the more natural
+  `EquicontinuousOn F (⋃₀ 𝔖)`. This is in order to keep the most generality, as the first statement
+  is strictly weaker, but this may well be completely useless so we might want to change this in
+  the future.
+
+* In Bourbaki, the usual Arzela-Ascoli compactness theorem follows from a similar total boundedness
+  result. Here we go directly for the compactness result, which is the most useful in practice, but
+  this will be an easy addition/refactor if we ever need it.
```

Comment:
Could you clarify what you mean by "this" here: I presume you mean establishing the intermediate total boundedness result a la Bourbaki?

Reply:
Don't you want me to change this before merging? I was planning to do it tomorrow morning. I can also change the wording later if you want.

Reply:
I decided I was splitting hairs and it wasn't worth delaying this PR any further.

Feel free to follow up with a doc string adjustment PR or not, according to your taste!